Refactor charmlite to avoid cyclic dependencies between headers

NK-Nikunj opened the pull request:
This PR provides the following changes:
1. Decouple common.hpp such that types instantiation of incomplete
types now occur in their respective headers.
2. Split message.hpp to message and impl. The implementation
relies on destination and the respective free functions are
defined there as well.
3. Split proxy.hpp to proxy and impl. The implementation of some
functions introduced cyclic dependency with completion and
collection.
4. Core related functions have been transferred to charmlite.hpp
and so has the functions from charmlite.impl.hpp.
5. Introduce Singletons as opposed to Cvs macros to keep single instance across multiple compilation units.
6. Enable warnings and debug options in Makefiles.
7. Decouple various Cvs macros by placing them underneath fully declared class types.

Random Flyby:
1. Change `enum` to `enum class`
2. Implement meta-template solution to `is_message`
3. Replace global variables to `static constexpr` in `struct`

NK-Nikunj committed:
Refactor charmlite to avoid cyclic dependencies between headers

This commit provides the following changes:
1. Decouple common.hpp such that types instantiation of incomplete
types now occur in their respective headers.
2. Split message.hpp to message and impl. The implementation
relies on destination and the respective free functions are
defined there as well.
3. Split proxy.hpp to proxy and impl. The implementation of some
functions introduced cyclic dependency with completion and
collection.
4. Core related functions have been transferred to charmlite.hpp
and so has the functions from charmlite.impl.hpp.

Random Flyby:
1. Change enum to enum class
2. Implement meta-template solution to is_message

jszaday commented on common.mk:
```diff
@@ -2,7 +2,7 @@
 CXX=$(CHARM_HOME)/bin/charmc -language converse++
 
-OPTS?=-g3
+OPTS?=-g -O0 -Wall -Wextra
```
Probably fine to include the `-W` flags, but `-O0` is a bit too extreme for CharmLite! Favor `-Og` at worst, maintain `-O2` for the general case, and `-Ofast` for benchmarks :)

NK-Nikunj replied:
That was mostly to limit any optimization from the compiler's end while I debugged. I'll change that back to `-g3` in the next commit.

jszaday replied:
Ah, note, `-Og` is generally optimal for debugging. Per the GCC manual, "some compiler passes that collect debug information are disabled at `-O0`."

jszaday commented on include/charmlite/core/common.hpp:
```diff
@@ -115,97 +123,48 @@ namespace cmk {
-    using callback_id_t = typename callback_table_t::size_type;
+    enum class destination_kind : std::uint8_t
+    {
+        kInvalid = 0,
```
We can probably the `k` prefix since these are enclosed within a class now. It's a holdover from the Google style guide, used to indicate global variables.

NK-Nikunj replied:
Sure, I can get rid of them.

jszaday commented on include/charmlite/charmlite.hpp:
```diff
@@ -1,21 +1,252 @@
+        register_function_<Message, callback_fn_t, Fn>(
+            CsvAccess(callback_table_));
+
+    inline completion* system_detector_(void)
```
Now that encapsulation is working correctly, `system_detector_` should be moved into _core.cpp_.

NK-Nikunj committed:
Replace Cvs macros to singletons

NK-Nikunj commented:
@jszaday I'll have to replace the CMK_ACCESS_SINGLETON macros with types instead of objects. Right now everything is glued together by another macro but I'll get back to it tomorrow.

jszaday commented on include/charmlite/core/common.hpp:
```diff
@@ -23,6 +23,35 @@ void CmiInitMemAffinity(char** argv);
+        singleton& operator=(singleton const&) = delete;
+        singleton& operator=(singleton&&) = delete;
+
+        static std::shared_ptr<T> instance()
```
It might be better to return a `const unique_ptr<T>&` here.
Reference counting will delay access in SMP scenarios and shared ownership isn't strictly necessary here.
Especially when we're obfuscating it with a macro.

NK-Nikunj replied:
I've added a CRTP based solution. Now the `CvsExtern` can be directly replaced with `CMK_GENERATE_SINGLETON` and `CvsAccess` with `CMK_ACCESS_SINGLETON` to achieve the exact same solution as the previous macros did (with Singleton behavior).

jszaday commented on include/charmlite/charmlite.hpp:
```diff
@@ -1,21 +1,253 @@
+    void exit(message_ptr<>&& msg);
+
+    template <entry_fn_t Fn, bool Constructor>
+    static entry_id_t register_entry_fn_(void)
```
Perhaps we could move all the `register_` functions (and their helpers) into _registration.hpp_?

NK-Nikunj replied:
Yeah, that's actually a better thing to do.

NK-Nikunj committed:
Review Changes

This commit updates the following:
1. Uses a CRTP based singleton solution as opposed to a generic
single-singleton per type solution. This allows multiple singleton
instances of the same type.
2. Add register.hpp which includes all the registration functions.
3. Put system_detector_ in core.cpp.
4. Update naming scheme of enum class.

NK-Nikunj committed:
Replace shared_ptr with unique_ptr in singleton
